include/cadmium/basic_model/iestream.hpp

```diff
@@ -0,0 +1,188 @@
+ *
+ * iestream PDEVS Model plays a history of events received by an input stream.
+ * The list of events allows to be used as a connector to external tools.
+ * The input format is "time output and a custom parser can be defined."
+ * Data type MSG must have the operator >> in order to work. 
+ * Data type TIME must also have the operator >> in order to work.
+ * Each line must be an MSG. Therefore the operator >> cannot read inputs that are specified in multiple lines
+ *
+*/
+
+
+template<class TIME, class INPUT>
+class Parser {
+    private:
+        std::ifstream file;
+    public:
+        // Constructors
+        Parser() {}
+        Parser(const char* file_path) {
+            this->open_file(file_path);
+        }
+        void open_file(const char* file_path) {
+            file.open(file_path);
+        }
+    std::pair<TIME,INPUT> next_timed_input() {
+        INPUT result;
+        TIME next_time;
+        if (file.eof()) throw std::exception();
+        file >> next_time;
+        file >> result;
+        return std::make_pair(next_time,result);
+    }
+};
+
+template<typename MSG> struct iestream_input_defs{
+    //custom ports
+    struct out : public out_port<MSG> {};
+};
+
+template<typename MSG, typename TIME>
+class iestream_input {
+    using defs=iestream_input_defs<MSG>;
+    public:
+        // default constructor
+        iestream_input() noexcept {}
+        iestream_input(const char* file_path) noexcept {
+            state._parser.open_file(file_path);
+        }
+    // state definition
+    struct state_type{
+        Parser<TIME, MSG> _parser;
+        MSG _last_input_read;
+        vector<MSG> _next_input;
+        TIME _simulation_time = TIME();
+        TIME _next_time = TIME();
+        TIME _next_time2 = TIME();
+        bool _initialization = true;
+    }; 
+    state_type state;
+
+    // ports definition
+    using input_ports=std::tuple<>;
+    using output_ports=std::tuple<typename defs::out>;
+
+    // internal transition
+    void internal_transition() {
+        //out << "INTERNAL. IESTREAM " << endl;
+        state._simulation_time += state._next_time;
+        state._next_input.clear();
+        if(state._initialization){
+            state._initialization = false;
+            try {
+                std::pair<TIME, MSG> parsed_line = state._parser.next_timed_input();
+                state._next_time = parsed_line.first - state._simulation_time;
+                if (state._next_time < TIME({0})) throw std::exception();
+                state._last_input_read = parsed_line.second;
+                state._next_input.push_back(state._last_input_read);
+            } catch(std::exception& e) {
+                state._next_time = std::numeric_limits<TIME>::infinity();
+            }
+            try {
+                std::pair<TIME, MSG> parsed_line = state._parser.next_timed_input();
+                state._next_time2 = parsed_line.first - state._simulation_time;
```

Comment:
can you make the name more explicit, what you mean by next_time2, it is easy to mix up with next_time.

Reply:
next_time2 is the next time after next_time